components/fxa-client/src/lib.rs

```diff
@@ -243,10 +254,28 @@ impl FirefoxAccount {
         }
     }
 
-    fn get_refresh_token(&self) -> Result<&str> {
+    fn get_refresh_token(&mut self) -> Result<&str> {
         match self.state.refresh_token {
             Some(ref token_info) => Ok(&token_info.token),
-            None => Err(ErrorKind::NoRefreshToken.into()),
+            None => {
+                log::info!("No refresh_token, attempting to recover via migration state");
+                // if we have no refresh token, check if this is the case of a failed migration
+                if self.state.in_flight_migration.is_some() {
+                    // if there's a pending migration state, then try to provision a refresh_token
+                    // by exchanging the session token for the refresh token
+                    self.helper_migration_network_methods();
```

Comment:
@rfk @eoger any smart ideas here how to deal with the mutable here? 

```
error[E0596]: cannot borrow `*self` as mutable, as it is behind a `&` reference
  --> components/fxa-client/src/device.rs:24:29
   |
23 |     pub fn get_devices(&self) -> Result<Vec<Device>> {
   |                        ----- help: consider changing this to be a mutable reference: `&mut self`
24 |         let refresh_token = self.get_refresh_token()?;
   |                             ^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow `*self` as mutable, as it is behind a `&` reference
   --> components/fxa-client/src/device.rs:104:29
    |
99  |         &self,
    |         ----- help: consider changing this to be a mutable reference: `&mut self`
...
104 |         let refresh_token = self.get_refresh_token()?;
    |                             ^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0502]: cannot borrow `self.client` as immutable because it is also borrowed as mutable
   --> components/fxa-client/src/device.rs:146:13
    |
144 |         let refresh_token = self.get_refresh_token()?;
    |                             ---- mutable borrow occurs here
145 |         let pending_commands =
146 |             self.client
    |             ^^^^^^^^^^^ immutable borrow occurs here
147 |                 .pending_commands(&self.state.config, refresh_token, index, limit)?;
    |                                                       ------------- mutable borrow later used here

error[E0502]: cannot borrow `self.state.config` as immutable because it is also borrowed as mutable
   --> components/fxa-client/src/device.rs:147:35
    |
144 |         let refresh_token = self.get_refresh_token()?;
    |                             ---- mutable borrow occurs here
...
147 |                 .pending_commands(&self.state.config, refresh_token, index, limit)?;
    |                                   ^^^^^^^^^^^^^^^^^^  ------------- mutable borrow later used here
    |                                   |
    |                                   immutable borrow occurs here

error[E0596]: cannot borrow `*self` as mutable, as it is behind a `&` reference
   --> components/fxa-client/src/device.rs:264:29
    |
263 |     fn update_device(&self, update: DeviceUpdateRequest<'_>) -> Result<UpdateDeviceResponse> {
    |                      ----- help: consider changing this to be a mutable reference: `&mut self`
264 |         let refresh_token = self.get_refresh_token()?;
    |                             ^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

```

Or do we need to rewrite all the cases where the refresh_token is used:

<img src=https://user-images.githubusercontent.com/128755/72536834-0708b280-3849-11ea-8173-c6886c31379e.png width=500 />

Reply:
No way around it, we are modifying the state.
You also need to keep in mind that these methods calling `refresh_token()` can now potentially mutate the persisted state, which means you need to call `tryPersistState` in the Kotlin/Swift wrappers after calling these. (also please add the `/// **💾 This method may alter the persisted account state.**` doc)